compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -655,7 +655,15 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
             else qual
         }
       else qual
-      typedSelect(tree, pt, qual1).withSpan(tree.span).computeNullable()
+      var sel = typedSelect(tree, pt, qual1).withSpan(tree.span).computeNullable()
+      if ctx.mode.is(Mode.UnsafeJavaReturn) && pt != AssignProto then
+        val sym = sel.symbol
+        if sym.is(JavaDefined) && sym.isTerm && !sym.is(Method) then
+          val stp1 = sel.tpe.widen
+          val stp2 = stp1.replaceOrNull
```

Comment:
Consider combining the `replaceOrNull`, `ne`, and `cast` functionality into a single helper method that can be reused both here for selections and above for applications. In addition to slightly reducing duplication, it would make the intent of `replaceOrNull` more clear.

Reply:
It is a little difficult to generalize this code, since we need to check if the type is changed to avoid extra assignment.